Construct custom query (#1348)
ocaml/ocaml-lsp: ocaml-lsp-server/docs/ocamllsp/construct-spec.md

# Construct Request

## Description

Provides commands to browse and fill typed holes (`_`). Such holes sometimes
appear in the result of other commands like `destruct` and can also be inserted
manually in the source. The command is already accessible via a completion hook,
however, in certain situations, invoking `construct` on a hole via a request
allows more control.

## Client Capability

There is no client capability relative to this request.

## Server capability

- property name: `handleConstruct`
- property type: `boolean`

## Request

> - method: `ocamllsp/typeEnclosing`

Reviewer: I assume typeEnclosing is a typo

Author: Fixed on https://github.com/ocaml/ocaml-lsp/pull/1348/commits/b2939c252136a00bd02e561a6275dd5f30ee1756

Nice catch, thanks!

- params:

  ```json
  {
    "uri": TextDocumentIdentifier,
    "position": Position
    "depth?": uinteger (default value: 0)
    "withValues?": <"local" | "none">,
  }
  ```

> 

Author: Fixed on https://github.com/ocaml/ocaml-lsp/pull/1348/commits/b2939c252136a00bd02e561a6275dd5f30ee1756

# Construct Request

## Description

> Provides commands to browse and fill typed holes (`_`). Such holes sometimes

Reviewer 2: ```suggestion
Provides commands to fill typed holes (`_`). Such holes sometimes
```

Author: Fixed on https://github.com/ocaml/ocaml-lsp/pull/1348/commits/b98322f1f1ed73cc9ad66b907b0c73c64c387448

appear in the result of other commands like `destruct` and can also be inserted
manually in the source. The command is already accessible via a completion hook,
however, in certain situations, invoking `construct` on a hole via a request
allows more control.

## Client Capability

There is no client capability relative to this request.

## Server capability

- property name: `handleConstruct`
- property type: `boolean`

## Request

- method: `ocamllsp/construct`
- params:

  ```json
  {
    "uri": TextDocumentIdentifier,
    "position": Position
    "depth?": uinteger (default value: 0)
    "withValues?": <"local" | "none">,
  }
  ```

The `depth` parameter allows to recursively construct terms. Note that
when `depth > 1` partial results of inferior depth will not be

> returned. The `withValues` use values from the environment (`local`)
> or not (`none`), by default, the parameter is fixed to `none`.

Reviewer 2: ```suggestion
returned. The `withValues` parameter enables the use of values from 
the environment (`local`) or not (`none`), It defaults to `none`.
```

Author: Fixed on https://github.com/ocaml/ocaml-lsp/pull/1348/commits/b98322f1f1ed73cc9ad66b907b0c73c64c387448
